File: src/rust/src/total_n_queens_test.rs
// https://leetcode.com/problems/n-queens-ii

pub fn solution(n: i32) -> i32 {
    fn placeQueens(row: i32, n: i32, mut leftDiagMap: i32, mut rightDiagMap: i32, mut colMap: i32) -> i32 {
        if row >= n {
            return 1;
        }

        let mut result = 0;
        for col in 0..n {
            let leftDiagShift = row + col;
            let rightDiagShift = n + (row - col);

            if (colMap & 1 << col) != 0
                || (leftDiagMap & 1 << leftDiagShift) != 0
                || (rightDiagMap & 1 << rightDiagShift) != 0 {
                continue;
            }

            colMap |= 1 << col;
            leftDiagMap |= 1 << leftDiagShift;
            rightDiagMap |= 1 << rightDiagShift;

            result += placeQueens(row + 1, n, leftDiagMap, rightDiagMap, colMap);

            colMap &= !(1 << col);
            leftDiagMap &= !(1 << leftDiagShift);
            rightDiagMap &= !(1 << rightDiagShift);
        }

        result
    }

    placeQueens(0, n, 0, 0, 0)
}

File: src/python/challenges/problems/total_n_queens_test.py
#https://leetcode.com/problems/n-queens-ii

from collections import defaultdict
from core.problem_base import *

class TotalNQueens(ProblemBase):
    def Solution(self, n: int) -> list[list[str]]:
        def placeQueens(row, leftDiagMap, rightDiagMap, colMap) -> int:
            if row >= n:
                return 1

            result = 0
            for col in range(n):
                leftDiagShift = row + col
                rightDiagShift = n + (row - col)

                if (colMap & 1 << col) != 0 \
                    or (leftDiagMap & 1 << leftDiagShift) != 0 \
                    or (rightDiagMap & 1 << rightDiagShift) != 0:
                    continue

                colMap |= 1 << col
                leftDiagMap |= 1 << leftDiagShift
                rightDiagMap |= 1 << rightDiagShift

                result += placeQueens(row + 1, leftDiagMap, rightDiagMap, colMap)

                colMap &= ~(1 << col)
                leftDiagMap &= ~(1 << leftDiagShift)
                rightDiagMap &= ~(1 << rightDiagShift)

            return result

        return placeQueens(0, 0, 0, 0)


if __name__ == '__main__':
    TestGen(TotalNQueens) \
        .Add(lambda tc: tc.Param(4).Result(2)) \
        .Add(lambda tc: tc.Param(1).Result(1)) \
        .Add(lambda tc: tc.Param(2).Result(0)) \
        .Add(lambda tc: tc.Param(3).Result(0)) \
        .Add(lambda tc: tc.Param(9).Result(352)) \
        .Run()
